### PR TITLE
GEOMESA-3278 Spark SQL - implement st_makevalid

### PR DESCRIPTION
**Pull request by thomas-maschler**
* implement st_makevalid using JTS GeometryFixer

### CODE REVIEW

File: geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/udf/GeometricProcessingFunctions.scala [hunk 1]
```diff
@@ -61,16 +61,23 @@ object GeometricProcessingFunctions {
     fastCircleToGeom(new JtsPoint(p, spatialContext).getBuffered(degrees, spatialContext))
   }
 
+  val ST_MakeValid: Geometry => Geometry = nullableUDF(geom => {
+    val fix = new util.GeometryFixer(geom)
+    fix.getResult
```
Comment (thomas-maschler) [inline geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/udf/GeometricProcessingFunctions.scala:66]: GeometryFixer allows to set two arguments
- isKeepCollapsed (defaults to `false`)
- isKeepMulti (defaults to `true`)

I decided to go with the default values to keep it simple.

File: geomesa-spark/geomesa-spark-jts/src/test/scala/org/locationtech/geomesa/spark/jts/udf/GeometricProcessingFunctionsTest.scala [hunk 2]
```diff
@@ -74,10 +74,54 @@ class GeometricProcessingFunctionsTest extends Specification with TestEnvironmen
...
+          "LINESTRING (0 0, 0 0)")
+        val expected_wkt = Seq("MULTIPOLYGON (((0 0, 0 1, 1 1, 1 0, 0 0)), ((1 1, 1 2, 2 2, 2 1, 1 1)))",
+          "MULTIPOLYGON (((149 80, 146 71, 142 62, 135 55, 128 48, 119 44, 110 41, 100 40, 90 41, 81 44, 72 48, 65 55, 64 56, 14 6, 13 6, 12 5, 11 5, 10 5, 9 5, 8 5, 7 6, 6 6, 6 7, 5 8, 5 9, 5 10, 5 11, 5 12, 6 13, 6 14, 56.76923076923077 64.76923076923077, 54 71, 51 80, 50 90, 51 100, 54 109, 58 118, 65 125, 72 132, 81 136, 90 139, 100 140, 110 139, 119 136, 125.23076923076923 133.23076923076923, 186 194, 187 194, 188 195, 189 195, 190 195, 191 195, 192 195, 193 194, 194 194, 194 193, 195 192, 195 191, 195 190, 195 189, 195 188, 194 187, 194 186, 134 126, 135 125, 142 118, 146 109, 149 100, 150 90, 149 80)))",
+          "MULTIPOLYGON (((51 10, 23 22, 11 50, 21.714285714285715 75, 11 100, 21.714285714285715 125, 11 150, 23 178, 51 190, 76 179.28571428571428, 101 190, 129 178, 141 150, 130.28571428571428 125, 141 100, 130.28571428571428 75, 141 50, 129 22, 101 10, 76 20.714285714285715, 51 10)))",
+          "LINESTRING EMPTY")
```
Comment (thomas-maschler) [inline geomesa-spark/geomesa-spark-jts/src/test/scala/org/locationtech/geomesa/spark/jts/udf/GeometricProcessingFunctionsTest.scala:112]: I took the test geometries from the [PostGIS documentation](https://postgis.net/docs/ST_MakeValid.html).
The results correspond to PostGIS `st_makeValid(geom, 'method=structure keepcollapsed=false')`

File: geomesa-spark/geomesa-spark-jts/src/test/scala/org/locationtech/geomesa/spark/jts/udf/GeometricProcessingFunctionsTest.scala [hunk 3]
```diff
@@ -74,10 +74,54 @@ class GeometricProcessingFunctionsTest extends Specification with TestEnvironmen
...
+        val expected = (false, true)
+
+        (decomposed.getBoolean(0), decomposed.getBoolean(1)) mustEqual expected
+        dfDecomposed mustEqual expected
+
+      }
+
+      "should repair geometry" >> {
+
+        val input_wkt = Seq("POLYGON ((0 0, 0 1, 2 1, 2 2, 1 2, 1 0, 0 0))",
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-jts/src/test/scala/org/locationtech/geomesa/spark/jts/udf/GeometricProcessingFunctionsTest.scala:105]:
```suggestion
        // these polygons are taken from the PostGIS docs: https://postgis.net/docs/ST_MakeValid.html
        val input_wkt = Seq("POLYGON ((0 0, 0 1, 2 1, 2 2, 1 2, 1 0, 0 0))", 
```

File: geomesa-spark/geomesa-spark-jts/src/test/scala/org/locationtech/geomesa/spark/jts/udf/GeometricProcessingFunctionsTest.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-spark/geomesa-spark-jts/src/test/scala/org/locationtech/geomesa/spark/jts/udf/GeometricProcessingFunctionsTest.scala:112]: So are the defaults you chose different than the PostGIS ones? (I don't actually see what the default is in the PostGIS docs there...) We should probably just document the differences. Could you add the new function to the docs [here](https://github.com/locationtech/geomesa/blob/main/docs/user/spark/sparksql_functions.rst), and just outline the default args and how they differ from the equivalent PostGIS function?

**Comment by elahrvivaz** [comment]
Thanks for the contribution! other than adding the function to the docs, it looks good to me

File: geomesa-spark/geomesa-spark-jts/src/test/scala/org/locationtech/geomesa/spark/jts/udf/GeometricProcessingFunctionsTest.scala (hunk 2 above)
Comment (thomas-maschler) [reply geomesa-spark/geomesa-spark-jts/src/test/scala/org/locationtech/geomesa/spark/jts/udf/GeometricProcessingFunctionsTest.scala:112]: The GeometryFixer works very similarly to the `structure` method from PostGis. The defaults I choose are in line with that method. The only differences I notice are that Polygons might start and end at different vertices, but the geometries line up nicely. 
PostGis introduced the `structure` method in version 3.2.0. It is currently not the default.

**Pull request merged**